bpo-29514: Check magic number for bugfix release

appeltel opened the pull request:
Add a test to check the current MAGIC_NUMBER against the
expected number for the release if the current release is
at candidate or final level. On test failure, describe to
the developer the procedure for changing the magic number.

https://bugs.python.org/issue29514

eliben commented:
Thanks for the contribution! This repository is currently a *read-only mirror*. For guidelines on contribution to CPython, please refer to [the devguide](https://docs.python.org/devguide/index.html)

appeltel committed:
bpo-29514: Check magic number for micro release

Add a dict importlib.util.EXPECTED_MAGIC_NUMBERS which
details the initial and expected pyc magic number for
each minor release. This gives a mechanism for users to
check if the magic number has changed within a release and
for a test to ensure procedure is followed if a change is
necessary.

Add a test to check the current MAGIC_NUMBER against the
expected number for the release if the current release is
at candidate or final level. On test failure, describe to
the developer the procedure for changing the magic number.

ncoghlan commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -248,6 +248,21 @@ def _write_atomic(path, data, mode=0o666):
+    (2, 7): 62211,
+    (3, 5): 3350,
+    (3, 6): 3379
+}
```
I'd put this table in the test suite rather than in the code - there's no need to have it other than the bytecode stability test case, and if it hasn't been updated when it needs to be we'll get a test failure anyway.

appeltel replied:
Simplifed to a sigle entry in the test case in 9ea7c52

ncoghlan commented on Lib/importlib/util.py:
```diff
@@ -4,6 +4,7 @@
 from ._bootstrap import _resolve_name
 from ._bootstrap import spec_from_loader
 from ._bootstrap import _find_spec
+from ._bootstrap_external import EXPECTED_MAGIC_NUMBERS
```
With the table moved to the test suite, this won't be needed any more.

appeltel replied:
removed in 9ea7c52

ncoghlan commented on Lib/test/test_importlib/test_release.py:
```diff
@@ -0,0 +1,57 @@
+
+        In exceptional cases, it may be required to change the MAGIC_NUMBER
+        for a maintenance release. In this case the change should be
+        discussed in dev-python. If a change is required, community
```
dev-python -> python-dev

appeltel replied:
fixed in 9ea7c52

ncoghlan reviewed (changes requested):
The general approach in the test case looks like it will give affected maintainers useful guidance, but the impact can be reduced by moving the `EXPECTED_MAGIC_NUMBERS` table into the test case itself (the test case will complain when it needs updating, so it's OK to have it separated from the list of actual magic numbers in the importlib code)

ambv commented on Lib/importlib/_bootstrap_external.py:
```diff
@@ -248,6 +248,21 @@ def _write_atomic(path, data, mode=0o666):
+
+EXPECTED_MAGIC_NUMBERS = {
+    (2, 7): 62211,
+    (3, 5): 3350,
```
Does this mean we're reverting the magic number bump from 3.5.3?

ncoghlan replied:
No, but the current special case from the test should just be moved directly into the table of expected magic numbers with a suitable comment.

appeltel committed:
Simplify magic number release test

Simplify the magic number release test by removing
EXPECTED_MAGIC_NUMBERS table and making the expected
magic number self-contained within the test.

BPO: 29514

ncoghlan commented on Lib/test/test_importlib/test_util.py:
```diff
@@ -757,5 +757,49 @@ def test_source_from_cache_path_like_arg(self):
  ) = util.test_both(PEP3147Tests, util=importlib_util)
 
 
+class MagicNumberTests:
```
Inherit from `unittest.TestCase` here

appeltel replied:
I initially avoided this and followed the test machinery as I wasn't sure about directly importing `importlib.util` within this test module, but after looking at the way the test specialization works I don't see how this is an issue. Changed in 1e32a1b.

ncoghlan replied:
Yeah, this module does some interesting things to create "real" test classes in different configurations, but we don't need those kinds of tricks for the new test.

ncoghlan commented on Lib/test/test_importlib/test_util.py:
```diff
@@ -757,5 +757,49 @@ def test_source_from_cache_path_like_arg(self):
+        adjustment to this test case.
+        """
+        if sys.version_info.releaselevel not in ('final', 'candidate'):
+            return
```
This can change to a `unittest.skipUnless` decorator on the testmethod:

    @unittest.skipUnless(sys.version_info.release in ('final', 'release'), 'Only check magic number stability for release candidates and later')

appeltel replied:
Done in 1e32a1b

ncoghlan commented on Lib/test/test_importlib/test_util.py:
```diff
@@ -757,5 +757,49 @@ def test_source_from_cache_path_like_arg(self):
+        actual = int.from_bytes(self.util.MAGIC_NUMBER[:2], 'little')
+
+        msg = (
+            "Candidate and final releases require the current  "
```
I'd start this message with "To avoid breaking backwards compatibility with cached bytecode files that can't be automatically regenerated by the current user, ..."

appeltel replied:
Added in 1e32a1b

ncoghlan reviewed (changes requested):
This is looking pretty close now, just a couple of suggestions around details of the test execution.

appeltel committed:
Improve magic number test execution and message

Improve the execution of the magic number test by
using skipUnless for alpha and beta releases, and
directly inheriting from unittest.TestCase rather than
using the machinery for the other tests. Also improve
the error message to explain the reason for caution in
changing the magic number.

BPO: 29514

serhiy-storchaka commented on Lib/test/test_importlib/test_util.py:
```diff
@@ -757,5 +758,48 @@ def test_source_from_cache_path_like_arg(self):
+        actual = int.from_bytes(importlib.util.MAGIC_NUMBER[:2], 'little')
+
+        msg = (
+            "To avoid breaking backwards compatibility with cached bytecode "
```
Wouldn't be better to use a multiline string literal if you want such verbose error message?

But I think that "broken *.pyc files compatibility" would be enough.

ambv replied:
Then you'd have to dedent it, etc. This message is perfectly legible to me.

serhiy-storchaka commented on Lib/test/test_importlib/test_util.py:
```diff
@@ -757,5 +758,48 @@ def test_source_from_cache_path_like_arg(self):
+    )
+    def test_magic_number(self):
+        """
+        Each python minor release should generally have a MAGIC_NUMBER
```
I think it would be better to write this as a comment rather than a docstring. If you want, add a one-line docstring, but most test methods don't have docstrings.

miss-islington commented:
🐍🍒⛏🤖 Thanks @appeltel for the PR, and @ambv for merging it 🌮🎉.I'm working now to backport this PR to: 2.7, 3.6.

miss-islington commented:
Sorry @appeltel and @ambv, I had trouble checking out the `3.6` backport branch.
                            Please backport using [cherry_picker](https://pypi.org/project/cherry-picker/) on command line.

miss-islington commented:
Sorry, @appeltel and @ambv, I could not cleanly backport this to `2.7` due to a conflict. 
                            Please backport using [cherry_picker](https://pypi.org/project/cherry-picker/) on command line.

ncoghlan commented:
This was backported a while back, but I missed updating the labels:

2.7:https://github.com/python/cpython/pull/2160 
3.5: https://github.com/python/cpython/pull/2158
3.6: https://github.com/python/cpython/pull/2157
